File: Backend/persistence/src/main/java/net/chrisparton/sparkled/persistence/song/impl/query/GetAllSongsQuery.java
package net.chrisparton.sparkled.persistence.song.impl.query;

import net.chrisparton.sparkled.model.entity.Song_;
import net.chrisparton.sparkled.model.entity.Song;
import net.chrisparton.sparkled.persistence.PersistenceQuery;

import javax.persistence.EntityManager;
import javax.persistence.TypedQuery;
import javax.persistence.criteria.CriteriaBuilder;
import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Root;
import java.util.List;

public class GetAllSongsQuery implements PersistenceQuery<List<Song>> {

    @Override
    public List<Song> perform(EntityManager entityManager) {
        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
        CriteriaQuery<Song> cq = cb.createQuery(Song.class);
        Root<Song> song = cq.from(Song.class);

        cq.orderBy(
                cb.asc(song.get(Song_.artist)),
                cb.asc(song.get(Song_.album)),
                cb.asc(song.get(Song_.name))
        );

        TypedQuery<Song> query = entityManager.createQuery(cq);
        return query.getResultList();
    }
}

File: persistence/src/main/kotlin/io/sparkled/persistence/song/impl/query/GetAllSongsQuery.kt
package io.sparkled.persistence.song.impl.query

import io.sparkled.model.entity.QSong.Companion.song
import io.sparkled.model.entity.Song
import io.sparkled.persistence.PersistenceQuery
import io.sparkled.persistence.QueryFactory

class GetAllSongsQuery : PersistenceQuery<List<Song>> {

    override fun perform(queryFactory: QueryFactory): List<Song> {
        return queryFactory
            .selectFrom(song)
            .orderBy(
                song.name.asc(),
                song.album.asc(),
                song.artist.asc()
            )
            .fetch()
    }
}
